Add example for parallel sort

Connor-GH opened the pull request:
It is now clearer how to use this library. The example(s) are built by default.

Commit:
Add example for parallel sort

It is now clearer how to use this library. The example(s) are built by default.

Commit:
Fix rustfmt issues from CI

hkaiser commented:
@Connor-GH thanks for looking into this. @SAtacker would you mind haveing a look?

pingu-73 commented:
@SAtacker @hkaiser  do we need to explicitly add examples 'coz it's pretty clear from [tests](https://github.com/STEllAR-GROUP/hpx-rs/blob/main/hpx-sys/src/lib.rs) how to use this library they simply work as examples.

pingu-73 commented:
@Connor-GH are you building hpx-rs on `linux` or `mac` ?

Connor-GH commented:
I build it on linux.

pingu-73 commented:
> I build it on linux.

could you please add instructions to build it on linux in the Readme.
Thanks

Connor-GH commented:
> @SAtacker @hkaiser do we need to explicitly add examples 'coz it's pretty clear from [tests](https://github.com/STEllAR-GROUP/hpx-rs/blob/main/hpx-sys/src/lib.rs) how to use this library they simply work as examples.

HPX has its tests, and it has its examples. Tests are meant to be entirely functional. People aren't really supposed to look at them until a regression pops up. On the other hand, HPX's examples are fairly small (can fit on one screen) and have some important comments. The examples are there for people who just need to quickly pick something up with the library and need to learn by doing, rather than searching for a specific flag or something (which is what the documentation is for). 

This is probably best demonstrated visibly. Here is a simple example from HPX that I learned from to help develop this example (notice the comments):
- https://github.com/STEllAR-GROUP/hpx/blob/master/examples/throttle/spin.cpp
If I were learning purely from tests, I would have to learn from code like this (notice the lack of comments):
- https://github.com/STEllAR-GROUP/hpx/blob/master/tests/regressions/block_matrix/algorithms.cpp

It should also be noted that examples should be separated from the implementation, but I don't think this matters wrt the testcases. Many languages have built-in unit tests, and I think putting them in the implementation file is fine. D does this for example in its standard library, but their tests are a bit more trivial but express the power of the API. There is a way of writing tests that can work as examples, and D does this this well, but these tests just serve to "get the thing done" with low regard for the potential of it being used to educate.

Maybe I am [bikeshedding](https://en.wikipedia.org/wiki/Law_of_triviality) here, but there is a genuine case for both examples and tests.

pingu-73 commented:
> > @SAtacker @hkaiser do we need to explicitly add examples 'coz it's pretty clear from [tests](https://github.com/STEllAR-GROUP/hpx-rs/blob/main/hpx-sys/src/lib.rs) how to use this library they simply work as examples.
> 
> HPX has its tests, and it has its examples. Tests are meant to be entirely functional. People aren't really supposed to look at them until a regression pops up. On the other hand, HPX's examples are fairly small (can fit on one screen) and have some important comments. The examples are there for people who just need to quickly pick something up with the library and need to learn by doing, rather than searching for a specific flag or something (which is what the documentation is for).
> 
> This is probably best demonstrated visibly. Here is a simple example from HPX that I learned from to help develop this example (notice the comments):
> 
> * https://github.com/STEllAR-GROUP/hpx/blob/master/examples/throttle/spin.cpp
>   If I were learning purely from tests, I would have to learn from code like this (notice the lack of comments):
> * https://github.com/STEllAR-GROUP/hpx/blob/master/tests/regressions/block_matrix/algorithms.cpp
> 
> Maybe I am [bikeshedding](https://en.wikipedia.org/wiki/Law_of_triviality) here, but there is a genuine case for both examples and tests.

i agree that examples are essential, but i don't think it's ideal to write one-liner examples as they don't add much value beyond what's already covered in the tests. It would be more beneficial to create more complex and practical examples. Since most of the functions maintain a similar structure and usage style as HPX it's ideal to refer to hpx documentation for understanding about that function (till the time we aren't done with hpx-rs documentation), more elaborate examples would help demonstrate their effective use.

@SAtacker what are your thoughts on it?

SAtacker commented on hpx-sys/src/lib.rs:
```diff
@@ -34,6 +34,7 @@ pub mod ffi {
         fn hpx_partial_sort_comp(src: &mut Vec<i32>, last: usize, comp: fn(i32, i32) -> bool);
     }
 }
+pub use self::ffi::*;
```
Hello,
Could you please explain this change?

pingu-73 replied:
for ffi bindings, it's generally not considered good practice to use `pub use self::ffi::*` because you want to maintain a clear separation between the ffi fun and the root of the crate. 
this helps in distinguishing between `unsafe/raw ffi calls` and any safe abstractions built on top of them. hence, there's often a separate crate for ffi bindings(hpx-sys) which is then used by a higher-level, more Rust-idiomatic crate(hpx-rs)

SAtacker replied:
@Connor-GH could you please remove this as @pingu-73 has described.

Connor-GH replied:
Okay, after doing more research, ffi as a namespace seems pretty common, as it is used in glib, gtk, and cairo. My original intention was to give the examples a more "seamless" feel, but it should not be done this way. My next proposal is that more ffi wrappers be written in the hpx_sys namespace so that the ffi functions don't show up in code as often. Example: create_c_args instead takes Args and does the conversion in the library. Alternatively, hpx_sys::ffi::init() could be replaced in user code with an hpx_sys::init that takes simply a function and Args.

SAtacker commented:
> > > @SAtacker @hkaiser do we need to explicitly add examples 'coz it's pretty clear from [tests](https://github.com/STEllAR-GROUP/hpx-rs/blob/main/hpx-sys/src/lib.rs) how to use this library they simply work as examples.
> > 
> > 
> > HPX has its tests, and it has its examples. Tests are meant to be entirely functional. People aren't really supposed to look at them until a regression pops up. On the other hand, HPX's examples are fairly small (can fit on one screen) and have some important comments. The examples are there for people who just need to quickly pick something up with the library and need to learn by doing, rather than searching for a specific flag or something (which is what the documentation is for).
> > This is probably best demonstrated visibly. Here is a simple example from HPX that I learned from to help develop this example (notice the comments):
> > 
> > * https://github.com/STEllAR-GROUP/hpx/blob/master/examples/throttle/spin.cpp
> >   If I were learning purely from tests, I would have to learn from code like this (notice the lack of comments):
> > * https://github.com/STEllAR-GROUP/hpx/blob/master/tests/regressions/block_matrix/algorithms.cpp
> > 
> > Maybe I am [bikeshedding](https://en.wikipedia.org/wiki/Law_of_triviality) here, but there is a genuine case for both examples and tests.
> 
> i agree that examples are essential, but i don't think it's ideal to write one-liner examples as they don't add much value beyond what's already covered in the tests. It would be more beneficial to create more complex and practical examples. Since most of the functions maintain a similar structure and usage style as HPX it's ideal to refer to hpx documentation for understanding about that function (till the time we aren't done with hpx-rs documentation), more elaborate examples would help demonstrate their effective use.
> 
> @SAtacker what are your thoughts on it?

I agree that trivial, one-liner examples may not add much value at this stage, as they are largely covered by the existing tests. 
It would be more beneficial to focus on more complex and practical examples that showcase real-world use cases, especially considering that most of the functions in hpx-rs follow similar patterns and usage as in HPX. 
For now, referring to HPX documentation for detailed function usage (until the HPX-RS documentation is ready) makes sense. 
That said, since there’s already an example in this PR, I have no objections to including it for now.

Commit:
Add more Rust-like hpx_sys::init() function that wraps hpx_sys::ffi::init()

Having to do all of the Args -> Vec<String> -> *mut *mut c_char conversions in every program using hpx-rs seems tedious. This wrapper helps to remove some of these ugly layers by moving them to the library rather than the user code

Connor-GH commented:
I apologize, I have been busy lately. I finally have a commit ready again for review though. By the time this PR is done, there should be no code that directly calls the ffi. IOW: all code needs to go through the "[front door](https://dlang.org/blog/2023/10/02/crafting-self-evident-code-with-d/)" of hpx_sys::, rather than hpx_sys::ffi::. 

Things I don't like about this code that I am asking guidance on:

- I would *really* prefer not to pass this function pointer through by another static variable, but it was the only way I could find to keep the API of fn(Vec<String>) to the user. It feels really bad, please work with me on fixing this.
- A bit too much unsafe code in the string conversion process and whatnot. Is there a safe way to go from Vec<String> to \*mut \*mut c_char? It seems like there would be.

Commit:
Regression: move unsafe{} block to cover the full assignment. The unsafe part of this assignment is itself, not the right-hand side.

pingu-73 commented on hpx-sys/src/lib.rs:
```diff
@@ -39,9 +40,45 @@ pub mod ffi {
 // Wrapper for the above Bindings.
 // reffer to tests to understand how to use them. [NOTE: Not all bindings have wrapper.]
 // ================================================================================================
-use std::ffi::CString;
```
@Connor-GH can you explain why have you removed `std::ffi::String`?
> I would really prefer not to pass this function pointer through by another static variable, but it was the only way I could find to keep the API of fn(Vec) to the user. It feels really bad, please work with me on fixing this.

- `CString` in Rust is a better approach, and it eliminate the need for the c_init function and the static variable.
- or ig instead of `static variable` you can use `closure` like the following
```
pub fn init<F>(func: F, func_args: Vec<String>) -> i32
where
    F: Fn(Vec<String>) -> i32 + 'static,
{
    let c_init = ......
    ....
    ...
}
```

Connor-GH replied:
> can you explain why have you removed `std::ffi::String`?
It is right here: https://github.com/STEllAR-GROUP/hpx-rs/pull/15/files#diff-7ab105a9531556e61127ea7119050c01f680c4d93e2fbaa66cb1a4c19bbe66f7R44

Connor-GH replied:
Please explain what you mean by "you can use ``closure``". A closure cannot be cast into a function pointer if it takes any arguments.

pingu-73 commented on hpx-sys/src/lib.rs:
```diff
@@ -39,9 +40,45 @@ pub mod ffi {
+use std::ffi::{CString, CStr};
 use std::os::raw::c_char;
 
+static mut FUNC_HOLDER: Option<fn (Vec<String>) -> i32> = None;
+
+// Convert arguments from *mut *mut c_char to Vec<String>
+// and call the saved Rust version of the function.
+fn c_init(argc: i32, argv: *mut *mut c_char) -> i32 {
+    let mut vec_args: Vec<String> = Vec::new();
+
+    for i in 0..argc {
+        // SAFETY: We get the argv from a conversion from Vec<String>.
+        // The conversion, which is safe, perserves the arguments.
+        // Therefore, no null pointers.
+
+        let c_str: &CStr = unsafe { CStr::from_ptr(*argv.offset(i as isize)) };
+        let string = c_str.to_string_lossy().into_owned();
+        vec_args.push(string.to_string());
+    }
+
+    unsafe { FUNC_HOLDER.unwrap()(vec_args) }
+}
+
+pub fn init(func: fn(Vec<String>) -> i32, func_args: Vec<String>) -> i32
+{
+    unsafe { FUNC_HOLDER = Some(func) };
+    let str_args: Vec<&str> = func_args
+        .iter()
+        .map(|s| s.as_str())
+        .collect::<Vec<&str>>();
+    let (argc, rust_argv) = create_c_args(&str_args);
+    let argv = rust_argv.into_raw_parts().0;
+
+    unsafe {
+        return self::ffi::init(c_init, argc, argv);
+    }
+}
```
>Please explain what you mean by "you can use closure". A closure cannot be cast into a function pointer if it takes any arguments.

remove the `static mut` variable and replace it with `boxed closure` to store callback. use `lifetime` with closure because it needs to live until explicitly freed.
refer(not boxed in it): https://github.com/STEllAR-GROUP/hpx-rs/pull/15#discussion_r1803658147

SAtacker replied:
Do we really need the `FUNC_HOLDER`? Can we not be a bit more barbaric and construct `vec_args` within here? This makes the `FUNC_HOLDER` go away at no cost.

The need to have a static variable global to this file is more of a burden to verify with the best practices and I intend to just go around it until I know better.

Connor-GH replied:
You cannot put the ``c_init`` function inside of the rust ``init`` function because it captures the dynamic environment (the Vec<String> function pointer). You cannot use a closure because that's not what the FFI allows. So the only way to do this is as follows (in pseudocode):

```
fn some_rust_callback(argc: i32, argv: *mut *mut c_char) -> i32 {
let args = c_args_to_vec(argc, argv);
return rust_func(args); // this is the problem! how do we get this function in scope?
}

fn rust_init(rust_func: fn(Vec<String>) -> i32, args: Vec<String>) -> i32 {
let (argc, argv) = vec_to_c_args(args);
return self::ffi::init(some_rust_callback, argc, argv);
}
```

Connor-GH commented:
How would we feel about this implementation?

https://github.com/Actyx/Actyx/blob/1f4bd6699a9345bde630248294b10bbc902d7230/rust/actyx/node-ffi/src/lib.rs#L70-L143

SAtacker commented:
> How would we feel about this implementation?
> 
> https://github.com/Actyx/Actyx/blob/1f4bd6699a9345bde630248294b10bbc902d7230/rust/actyx/node-ffi/src/lib.rs#L70-L143

I lost the context, could you please remind?

SAtacker commented on hpx-sys/src/lib.rs:
```diff
@@ -1,6 +1,7 @@
 #![doc(html_root_url = "https://github.com/STEllAR-GROUP/hpx-rs")]
 #![allow(bad_style, non_camel_case_types, unused_extern_crates)]
 #![allow(dead_code, unused_imports)]
+#![feature(vec_into_raw_parts)]
```
The CI complains about this. PTAL

Connor-GH replied:
Are we committed to using only safe Rust?

Connor-GH replied:
Actually, I believe that it doesn't make sense to change to unstable Rust so fast unless we *need* it.

SAtacker replied:
The question is if we can keep using stable Rust to compile and continue without the nightly version needed to compile the code, how much more efforts would it need?

SAtacker replied:
I agree with you. Let's fix that before merging.

Connor-GH replied:
In that case, we do not have access to std:random. for that example, we will need to pull in the rand crate. Is this acceptable?

SAtacker replied:
>  Is this acceptable?

I would hope so unless there is something I don't know.

Connor-GH replied:
Dependency creep, basically.

Commit:
Remove code that required nightly

Commit:
Fix formatting for rustfmt CI

pingu-73 commented:
@SAtacker LGTM
